File: code/search/interpolation_search/interpolation_search.go
package main

import "fmt"

func interpolationSearch(array []int, key int) int {

	min, max := array[0], array[len(array)-1]

	low, high := 0, len(array)-1

	for {
		if key < min {
			return low
		}

		if key > max {
			return high + 1
		}

		var guess int
		if high == low {
			guess = high
		} else {
			size := high - low
			offset := int(float64(size-1) * (float64(key-min) / float64(max-min)))
			guess = low + offset
		}

		if array[guess] == key {
			for guess > 0 && array[guess-1] == key {
				guess--
			}
			return guess
		}

		if array[guess] > key {
			high = guess - 1
			max = array[high]
		} else {
			low = guess + 1
			min = array[low]
		}
	}
}

func main() {
	items := []int{1, 2, 9, 20, 31, 45, 63, 70, 100}
	fmt.Println(interpolationSearch(items, 63))
}

File: code/search/src/interpolation_search/interpolation_search.py
def interpolation_search(arr, n, x):
    low = 0
    high = (n - 1)

    while low <= high and x >= arr[low] and x <= arr[high]:
        mid = low + int(
            ((float(high - low) / (arr[high] - arr[low])) * (x - arr[low])))

        if arr[mid] == x:
            return mid

        if arr[mid] < x:
            low = mid + 1

        else:
            high = mid - 1

    return -1
